Added Machine ID CLI and configuration references. (#11096)
gravitational/teleport: docs/pages/machine-id/reference/cli.mdx

| `-a/--auth-server`   | Address of the Teleport Auth Server (On-Prem installs) or Proxy Server (Cloud installs).       |
| `--token`            | A bot join token, if attempting to onboard a new bot; used on first connect.                   |
| `--ca-pin`           | CA pin to validate the auth server.                                                            |
| `--data-dir`         | Directory to store internal bot data. Access to this directory should be limited.              | 
| `--destination-dir`  | Directory to write short-lived machine certificates.                                           |
| `--certificate-ttl`  | TTL of short-lived machine certificates.                                                       |
| `--renewal-interval` | Interval at which short-lived certificates are renewed; must be less than the certificate TTL. |
| `--join-method`      | Method to use to join the cluster, can be `token` or `iam`.                                    |
| `--one-shot`         | If set, quit after the first renewal.                                                          |

## `tbot init`

> Initialize and configure Linux ACLs for certificate destination directory. Used

Reviewer: ```suggestion
Initialize and configure Linux ACLs for the certificate destination directory. Used
```

> to lock down access to certificates on host when running long lived Machine ID

Reviewer: ```suggestion
to lock down access to certificates on a host when running a long-lived Machine ID
```

service.

```code
$ tbot init \
    --destination-dir=/opt/machine-id \
    --bot-user=ubuntu \
    --owner=ubuntu:ubuntu \
    --reader-user=jenkins
```

| Flag                 | Description                                                                                           |
|----------------------|-------------------------------------------------------------------------------------------------------|
| `-d/--debug`         | Enable verbose logging to stderr.                                                                     |
| `-c/--config`        | Path to a configuration file. Defaults to `/etc/tbot.yaml` if unspecified.                            |
| `--destination-dir`  | Directory to write short-lived machine certificates to.                                               |

> | `--owner`            | Defines Linux user:group owner of `--destination-dir`. Defaults to the bot user if unspecified.       |

Reviewer: ```suggestion
| `--owner`            | Defines Linux `user:group` owner of `--destination-dir`. Defaults to the bot user if unspecified.       |
```

---
title: Machine ID CLI Reference
description: CLI reference for Teleport Machine ID.
---

> # Machine ID CLI Reference

Reviewer 2: I don't think we need H1 here, because `title` already generates H2

## `tbot start`

Starts the Machine ID client `tbot`, fetching and writing certificates to disk
at a set interval.

<Details scope={["cloud"]} scopeOnly={true}>
```code
$ tbot start \

>    --data-dir=/var/lib/teleport/bot \

Reviewer 3: Maybe a nitpick, but this is the default value for `--data-dir` so this doesn't strictly need to be specified.

```code
$ tbot start \
   --data-dir=/var/lib/teleport/bot \
   --destination-dir=/opt/machine-id \
   --token=00000000000000000000000000000000 \
   --join-method=token \
   --ca-pin=sha256:1111111111111111111111111111111111111111111111111111111111111111 \
   --auth-server=auth.example.com:3025
```
</Details>

| Flag                 | Description                                                                                    |
|----------------------|------------------------------------------------------------------------------------------------|
| `-d/--debug`         | Enable verbose logging to stderr.                                                              |
| `-c/--config`        | Path to a configuration file. Defaults to `/etc/tbot.yaml` if unspecified.                     |
| `-a/--auth-server`   | Address of the Teleport Auth Server (On-Prem installs) or Proxy Server (Cloud installs).       |
| `--token`            | A bot join token, if attempting to onboard a new bot; used on first connect.                   |
| `--ca-pin`           | CA pin to validate the auth server.                                                            |

> | `--data-dir`         | Directory to store internal bot data. Access should be limited. See `tbot init` for details.   |

Reviewer 3: `tbot init` doesn't (currently at least) touch the internal data dir, only destinations, so this might be a bit of a red herring.

| `--destination-dir`  | Directory to write short-lived machine certificates.                                           |
| `--certificate-ttl`  | TTL of short-lived machine certificates.                                                       |
| `--renewal-interval` | Interval at which short-lived certificates are renewed; must be less than the certificate TTL. |
| `--join-method`      | Method to use to join the cluster, can be `token` or `iam`.                                    |
| `--one-shot`         | If set, quit after the first renewal.                                                          |

## `tbot init`

Initialize a certificate destination directory for writes from a separate bot

> user. Useful when running the bot under an isolated user account.

Author: Tim: Can you provide some more details on `tbot init`, specifically around how it can be used with Linux ACLs? Basically we want to explain to the user when/why they would want to use `tbot init`.

```code
$ tbot init \
    --destination-dir=/opt/machine-id \
    --bot-user=ubuntu \
    --owner=ubuntu:ubuntu \
    --reader-user=jenkins
```

| Flag                 | Description                                                                                           |
|----------------------|-------------------------------------------------------------------------------------------------------|
| `-d/--debug`         | Enable verbose logging to stderr.                                                                     |
| `-c/--config`        | Path to a configuration file. Defaults to `/etc/tbot.yaml` if unspecified.                            |
| `--destination-dir`  | Directory to write short-lived machine certificates to.                                               |

> | `--owner`            | Defines Linux user:group owner of `--destination-dir`. Defaults to `nobody:nobody` if unspecified.    |

Reviewer 3: Mentioned again below, but this string is no longer accurate. PR to fix that: https://github.com/gravitational/teleport/pull/11158

```suggestion
| `--owner`            | Defines Linux user:group owner of `--destination-dir`. Defaults to the bot user if unspecified.       |
```

| `--bot-user`         | When using ACLs, defines Linux user that can write short-lived certificates to `--destination-dir`.   |

> | `--reader-user`      | When using ACLs, defines Linux user that will read short-lived certificates from `--destination-dir`. |

Author: Tim: For these three (`--owner`, `--bot-user`, and `--reader-user`), can you provide some more details on how they should be used.

For example, one thing that is not clear to me is what flags trigger the use of Linux ACLs?

Reviewer 3: It will try to use ACLs if both `--bot-user` and `--reader-user` are set (and by default produces a loud warning if one or both are missing). I guess importantly, that's only on Linux, ACLs are never tried on macOS / Windows.

`init` tries to infer the owner automatically. That help message is outdated ([oops](https://github.com/gravitational/teleport/pull/11158)) as the default owner is now the bot user, but you can pass in `--owner` to use any arbitrary user:group. It'll try to fix ownership if possible (i.e. running `tbot init` as root) regardless of whether or not ACLs are in use.

Author: For `--owner` do you mean bot user or current Linux user that is running `tbot init`?

> | `--init-dir`         | If using a config file and multiple destinations are configured, specify which to initialize.         |

Author: Tim: I copied this over from the CLI flag help, but to be honest, I didn't quite understand the purpose of this flag.

Reviewer 3: Yeah, it's a bit vague, especially since we also have `--destination-dir`. The two flags control how we handle configuration:
 - `--destination-dir` adds an otherwise-unconfigured destination dir just like `tbot start` when there's no config file, potentially overriding destinations if they exist in your config file.
 - `--init-dir` just selects which directory to initialize from your config file without potentially altering your config. This is optional if using a config file with only 1 destination configured, since it will infer it automatically.

Admittedly having this split across two flags is a bit of a kludge. I'd like to try merging them together at some point but there were some hiccups integrating that with our config parsing.

| `-a/--auth-server`   | Address of the Teleport Auth Server (On-Prem installs) or Proxy Server (Cloud installs).       |
| `--token`            | A bot join token, if attempting to onboard a new bot; used on first connect.                   |
| `--ca-pin`           | CA pin to validate the Teleport Auth Server; used on first connect.                            |
| `--data-dir`         | Directory to store internal bot data. Access to this directory should be limited.              | 
| `--destination-dir`  | Directory to write short-lived machine certificates.                                           |
| `--certificate-ttl`  | TTL of short-lived machine certificates.                                                       |
| `--renewal-interval` | Interval at which short-lived certificates are renewed; must be less than the certificate TTL. |
| `--join-method`      | Method to use to join the cluster, can be `token` or `iam`.                                    |
| `--one-shot`         | If set, quit after the first renewal.                                                          |

## `tbot init`

> If you want to write certificates to disk as a different user than the Machine

Reviewer: It might be good to mention an example in this paragraph of when you would use this so a user scrolling down this page can have a better idea of whether this command is relevant for their use case.

ID client, you can use `tbot init` to configure either file or POSIX ACLs
permissions.

| Flag                | Description                                                                                                        |
|---------------------|--------------------------------------------------------------------------------------------------------------------|
| `-d/--debug`        | Enable verbose logging to stderr.                                                                                  |
| `-c/--config`       | Path to a configuration file. Defaults to `/etc/tbot.yaml` if unspecified.                                         |
| `--destination-dir` | Directory to write short-lived machine certificates to.                                                            |

> | `--owner`           | Defines Linux `user:group` owner of `--destination-dir`. Defaults to the Linux user running `tbot` if unspecified. |

Reviewer: ```suggestion
| `--owner`           | Defines the Linux `user:group` owner of `--destination-dir`. Defaults to the Linux user running `tbot` if unspecified. |
```

> | `--bot-user`        | Enables POSIX ACLs and defines Linux user that can read/write short-lived certificates to `--destination-dir`.     |

Reviewer: ```suggestion
| `--bot-user`        | Enables POSIX ACLs and defines the Linux user that can read/write short-lived certificates to `--destination-dir`.     |
```

> | `--reader-user`     | Enables POSIX ACLs and defines Linux user that will read short-lived certificates from `--destination-dir`.        |

Reviewer: ```suggestion
| `--reader-user`     | Enables POSIX ACLs and defines the Linux user that will read short-lived certificates from `--destination-dir`.        |
```

| `--token`            | A bot join token, if attempting to onboard a new bot; used on first connect.                   |
| `--ca-pin`           | CA pin to validate the Teleport Auth Server; used on first connect.                            |
| `--data-dir`         | Directory to store internal bot data. Access to this directory should be limited.              | 
| `--destination-dir`  | Directory to write short-lived machine certificates.                                           |
| `--certificate-ttl`  | TTL of short-lived machine certificates.                                                       |
| `--renewal-interval` | Interval at which short-lived certificates are renewed; must be less than the certificate TTL. |
| `--join-method`      | Method to use to join the cluster. Can be `token` or `iam`.                                    |
| `--one-shot`         | If set, quit after the first renewal.                                                          |

## `tbot init`

If you want to write certificates to disk as a different user than the Machine
ID client, you can use `tbot init` to configure either file or POSIX ACLs

> permissions. This allows you to lock down access to Machine ID short-lived

Reviewer: ```suggestion
permissions. This allows you to lock down access to Machine ID's short-lived
```
